### PR TITLE
Set up spack environment to handle build

### PR DESCRIPTION
**Pull request by cmacmackin**
# Description

The current build instructions are a bit ad-hoc. I have attempted to make the build process more robust using Spack. In particular, I have added a `spack.yaml` file which defines a custom [Spack environment](https://spack-tutorial.readthedocs.io/en/latest/tutorial_environments.html) to be [used for development](https://spack-tutorial.readthedocs.io/en/latest/tutorial_developer_workflows.html). 

Addressed (at least partially) #3, #20, relates to #86.

## Type of change

- [x] Build system improvements
- [x] Requires documentation updates

# Testing

The build was executed and confirmed to work on my laptop. The resulting test binaries which were built were run and all passed.

**Test Configuration**:

* OS: Linux Mint 21 (Ubuntu 22.04)
* SYCL implementation: HipSYCL 0.9.2, DPC++ 2022.1.0
* MPI details: MPICH 4.0.2
* Hardware: Intel Icelake

# Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any new dependencies are automatically built for users via `cmake`
- [x] I have used understandable variable names
- [x] I have run `clang-format` against my `*.hpp` and `*.cpp` changes
- [x] I have run `cmake-format` against my changes to `CMakeLists.txt`
- [x] I have run `black` against changes to `*.py`
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by cmacmackin** [comment]
@jwscook @will-saunders-ukaea This is in a state I'm reasonably happy with. It is ready for review and can be merged after NESO-Spack#3.

**Review by jwscook — CHANGES_REQUESTED** [review]
... now you can see my comments...

File: CMakeLists.txt [hunk 1]
```diff
@@ -55,7 +55,18 @@ message(STATUS "Found Nektar++: version ${NEKTAR++_VERSION}")
 # Set RPATH for wherever Nektar++ is installed.
 set(CMAKE_INSTALL_RPATH "${NEKTAR++_LIBRARY_DIRS}")
 
+# Hack to deal with the fact Intel seems to be inconsistent with whether it
+# prefers you to use dpcpp or icpx directly. MKL requires the compiler to be
+# dpcpp, but the DPCPP cmake files actually prefer that icpx is used.
+if(CMAKE_CXX_COMPILER MATCHES "icpx$")
+  message("REGEX evaluated true")
```
[the remark this answers is not in the document]
Comment (cmacmackin) [reply CMakeLists.txt:62]: Oh, that should just be removed, actually. It was just in there for debugging purposes.

File: spack.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# This is a Spack Environment file.
+#
+# It describes the set of packages to be installed when developing
+# NESO, along with configuration settings.
+spack:
+  # add package specs to the `specs` list
+  specs:
+  - neso%gcc ^openblas
```
[the remark this answers is not in the document]
Comment (cmacmackin) [reply spack.yaml]: I think this diff is out of date (not sure how the interface allowed that to happen). If you check the files changed tab now you'll see that `^hipsycl` is there.

File: README.md [hunk 3]
```diff
@@ -32,7 +134,7 @@ cd cmake-3.23.1/
 make
 ```
 
-### Boost
+#### Boost
 
 The test suite requires the [Boost library](https://www.boost.org/) (version >= 1.78).
```
[the remark this answers is not in the document]
Comment (cmacmackin) [reply README.md:139]: Updated (along with the dependency in the NESO package).

**Pull request merged**